dns/dnssecalgs/__init__.py

```diff
@@ -0,0 +1,106 @@
+from dns.rdtypes.ANY.DNSKEY import DNSKEY
+
+algorithms: Dict[Tuple[Algorithm, Optional[bytes]], Type[GenericPrivateKey]] = {
+    (Algorithm.RSAMD5, None): PrivateRSAMD5,
+    (Algorithm.DSA, None): PrivateDSA,
+    (Algorithm.RSASHA1, None): PrivateRSASHA1,
+    (Algorithm.DSANSEC3SHA1, None): PrivateDSANSEC3SHA1,
+    (Algorithm.RSASHA1NSEC3SHA1, None): PrivateRSASHA1NSEC3SHA1,
+    (Algorithm.RSASHA256, None): PrivateRSASHA256,
+    (Algorithm.RSASHA512, None): PrivateRSASHA512,
+    (Algorithm.ECDSAP256SHA256, None): PrivateECDSAP256SHA256,
+    (Algorithm.ECDSAP384SHA384, None): PrivateECDSAP384SHA384,
+    (Algorithm.ED25519, None): PrivateED25519,
+    (Algorithm.ED448, None): PrivateED448,
+}
+
+
+def get_algorithm_cls(
+    algorithm: Union[int, str], prefix: Optional[bytes] = None
+) -> Type[GenericPrivateKey]:
+    """Get Private Key class from Algorithm.
+
+    *algorithm*, a ``str`` or ``int`` specifying the DNSKEY algorithm.
+
+    Raises ``UnsupportedAlgorithm`` if the algorithm is unknown.
+
+    Returns a ``dns.dnssecalgs.GenericPrivateKey``
+    """
+    algorithm = Algorithm.make(algorithm)
+    cls = algorithms.get((algorithm, prefix))
+    if cls:
+        return cls
+    raise UnsupportedAlgorithm(
+        'algorithm "%s" not supported by dnspython' % Algorithm.to_text(algorithm)
+    )
+
+
+def get_algorithm_cls_from_dnskey(dnskey: DNSKEY) -> Type[GenericPrivateKey]:
+    """Get Private Key class from DNSKEY.
+
+    *dnskey*, a ``DNSKEY`` to get Algorithm class for.
+
+    Raises ``UnsupportedAlgorithm`` if the algorithm is unknown.
+
+    Returns a ``dns.dnssecalgs.GenericPrivateKey``
+    """
+    prefix = None
+    if dnskey.algorithm == Algorithm.PRIVATEDNS:
+        _, length = dns.name.from_wire(dnskey.key, 0)
+        prefix = dnskey.key[0:length]
+    elif dnskey.algorithm == Algorithm.PRIVATEOID:
+        length = int(dnskey.key[0])
+        prefix = dnskey.key[0 : length + 1]
+    return get_algorithm_cls(dnskey.algorithm, prefix)
+
+
+def register_algorithm_cls(
+    algorithm: Union[int, str],
+    algorithm_cls: Type[GenericPrivateKey],
+    name: Optional[Union[dns.name.Name, str]] = None,
+    oid: Optional[bytes] = None,
+) -> None:
+    """Register Algorithm Private Key class.
+
+    *algorithm*, a ``str`` or ``int`` specifying the DNSKEY algorithm.
+
+    *algorithm_cls*: A `GenericPrivateKey` class.
+
+    *name*, an optional ``dns.name.Name`` or ``str``, for for PRIVATEDNS algorithms.
+
+    *oid*: an optional BER-encoded `bytes` for PRIVATEOID algorithms.
+
+    Raises ``ValueError`` if a name or oid is specified incorrectly.
+    """
+    algorithm = Algorithm.make(algorithm)
+    if not issubclass(algorithm_cls, GenericPrivateKey):
+        raise TypeError("Invalid algorithm class")
+    prefix = None
+    if algorithm == Algorithm.PRIVATEDNS and name:
+        if isinstance(name, str):
+            name = dns.name.from_text(name)
+        prefix = name.to_wire()
```

Comment:
Do we need to worry about name canonicalization here, and in `get_algorithm_cls_from_dnskey`?  RFC 4034 says that the name used for the PRIVATEDNS algorithm must not be compressed in the DNSKEY rdata, but doesn't say anything about case.  As it's a domain name, I'd assume that case doesn't matter, which means that we'd either need to canonicalize on both insertion and lookup, or store the name in the map instead of the encoded bytes.

Reply:
Good point. Storing a downcased name in the map seems reasonable. Take a look at fa243d1163d3ad6f7f2033b837c7cdac19c6e8b1 for a possible solution. It's not actually a "prefix" anymore, more of a lookup key, so we might want to change the terminology.

Reply:
That commit appears to canonicalize the name and convert it to a string before storing it in the map.  Is there some reason not to store the `dns.name.Name`object directly (which wouldn't require canonicalization, since it's already case-insensitive)?

Reply:
No, there is no (good) reason - fixed in 0f7d05eca870461262957b8fb0a62941e99cc18a